File: erlang/memo.erl
%% memo.erl
%% Memoize a function that accepts a list of primitive args.
%% Primitives: integers, atoms (true/false), binaries, etc.
%% We use nested maps as a trie keyed directly by those terms.

-module(memo).
-export([memoize/1]).

-define(LEAF, '$leaf').

%% Public API:
%% Fun must accept a single argument: a list of args (e.g., [2,3,5]).
%% Returns a fun that takes that same list and caches by the list as a path.
memoize(Fun) when is_function(Fun, 1) ->
  ServerPid = spawn(fun() -> memo_server_loop(Fun, #{}) end),
  fun(ArgsList) when is_list(ArgsList) ->
      call_server(ServerPid, ArgsList)
  end.

%% Client: sync request/reply to the memo server
call_server(Pid, Args) ->
  Ref = make_ref(),
  Pid ! {lookup, self(), Ref, Args},
  receive
    {Ref, Value} -> Value
  end.

%% Memo server loop
memo_server_loop(Fun, Trie) ->
  receive
    {lookup, From, Ref, Args} ->
      case trie_lookup(Args, Trie) of
        {hit, Value} ->
          From ! {Ref, Value},
          memo_server_loop(Fun, Trie);
        miss ->
          Value = Fun(Args),
          NewTrie = trie_insert(Args, Value, Trie),
          From ! {Ref, Value},
          memo_server_loop(Fun, NewTrie)
      end
  end.

%% Trie operations (nested maps by arg path)
trie_lookup([], Trie) ->
  case maps:get(?LEAF, Trie, undefined) of
    undefined -> miss;
    V -> {hit, V}
  end;
trie_lookup([K | Rest], Trie) ->
  case maps:get(K, Trie, undefined) of
    undefined -> miss;
    SubTrie when is_map(SubTrie) -> trie_lookup(Rest, SubTrie);
    _ -> miss
  end.

trie_insert([], Value, Trie) ->
  Trie#{?LEAF => Value};
trie_insert([K | Rest], Value, Trie) ->
  Sub0 = maps:get(K, Trie, #{}),
  Sub1 = trie_insert(Rest, Value, Sub0),
  Trie#{K => Sub1}.

File: elixir/memo.ex
defmodule Memo do
  @moduledoc """
  Memoization using a trie (nested maps) keyed by a list of primitive args.
  Primitives: integers, booleans, atoms, binaries.

  Returns a closure that keeps its cache in an Agent process.
  Cache lifetime = Agent lifetime (ephemeral).
  """

  @leaf :leaf

  @doc """
  Returns a function that memoizes `fun`, which must accept a single list of args.

  Example:
      memo = Memo.memoize(&Enum.sum/1)
      memo.([1, 2, 3]) #=> 6
  """
  def memoize(fun) when is_function(fun, 1) do
    {:ok, agent} = Agent.start_link(fn -> %{} end)

    fn args_list when is_list(args_list) ->
      Agent.get_and_update(agent, fn trie ->
        case trie_lookup(args_list, trie) do
          {:hit, value} ->
            {value, trie}

          :miss ->
            value = fun.(args_list)
            {value, trie_insert(args_list, value, trie)}
        end
      end)
    end
  end

  # Trie ops
  defp trie_lookup([], trie) do
    case Map.get(trie, @leaf) do
      nil -> :miss
      v -> {:hit, v}
    end
  end

  defp trie_lookup([k | rest], trie) do
    case Map.get(trie, k) do
      nil -> :miss
      sub when is_map(sub) -> trie_lookup(rest, sub)
      _ -> :miss
    end
  end

  defp trie_insert([], value, trie), do: Map.put(trie, @leaf, value)

  defp trie_insert([k | rest], value, trie) do
    sub0 = Map.get(trie, k, %{})
    sub1 = trie_insert(rest, value, sub0)
    Map.put(trie, k, sub1)
  end
end
